### PR TITLE
Python Interface (SWIG)

### PR DESCRIPTION
**Pull request by kevinhkhuynh**
The compilation process should download and install SWIG 4.0.2 for you, but a few paths are hardcoded to my paths and need to be updated.

This is still in an unworking state.

Instructions while inside base directory of libROM:

python3 -m venv py3
source py3/bin/activate
pip install --upgrade pip
pip install numpy mpi4py h5py
./scripts/compile.sh -mp (Not working here after pulling in updates from master. Was working previously.)

### CODE REVIEW

File: lib/librom.i [hunk 1]
```diff
@@ -0,0 +1,93 @@
+%module librom
+
+//TODO: Untie this from LC.
+%include /usr/tce/packages/python/python-2.7.16/lib/python2.7/site-packages/mpi4py/include/mpi4py/mpi4py.i
```
Comment (kevinhkhuynh) [inline lib/librom.i:4]: This path must be updated. It is tied to LC.

Comment (kevinhkhuynh) [inline lib/librom.i:4]: This path must be updated. It is tied to LC.

File: scripts/compile.sh [hunk 2]
```diff
@@ -96,4 +102,18 @@ elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
...
+if [[ $INSTALL_PYTHON == "true" ]]; then
+    . ${REPO_PREFIX}/scripts/python_setup.sh
+    cd ${REPO_PREFIX}/lib
+    ${REPO_PREFIX}/dependencies/swig/swig_install/bin/swig -I${REPO_PREFIX}/dependencies/swig/swig_install/share/swig/4.0.2 -I${REPO_PREFIX}/dependencies/swig/swig_install/share/swig/4.0.2/python -I${REPO_PREFIX}/dependencies/swig/swig_install/share/swig/4.0.2/std -c++ -python -py3 librom.i
+    mv librom_wrap.cxx $LIB_BUILD_DIR
+    mv librom.py $LIB_BUILD_DIR
+    cd $LIB_BUILD_DIR
+
+    # TODO: Untie the Python lib from LC.
+    mpicxx -O2 -fPIC -DSWIG -c librom_wrap.cxx -I/usr/include/python3.6m -I/usr/workspace/huynh24/libROM4/env3/lib/python3.7/site-packages/mpi4py/include -I${REPO_PREFIX}/lib
```
Comment (kevinhkhuynh) [inline scripts/compile.sh:117]: This path must be updated. It is tied to my username.

File: scripts/compile.sh [hunk 3]
```diff
@@ -96,4 +102,18 @@ elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
...
+    mpicxx -shared -L/usr/workspace/huynh24/libROM4/env3/lib/python3.7/site-packages -Wl,-rpath,$LIB_BUILD_DIR -L$LIB_BUILD_DIR librom_wrap.o -lmkl_scalapack_lp64 -lmkl_blacs_intelmpi_lp64 -liomp5 -lpthread -lm -ldl -lROM -o _librom.so
```
Comment (kevinhkhuynh) [inline scripts/compile.sh:118]: This path must be updated. It is tied to my username.

File: lib/librom.i [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+%pointer_functions(int, intp);
+
+/* Create some functions for working with "double *" */
+%pointer_functions(double, doublep);
+
+%{
+    #define SWIG_FILE_WITH_INIT
+    #include "utils/CSVDatabase.h"
+    #include "utils/HDFDatabase.h"
+    #include "linalg/Vector.h"
```
Comment (kevinhkhuynh) [inline lib/librom.i:26]: Some files have not been included as I have not been able to get them to work yet, such as the greedy algorithm files.

File: lib/algo/greedy/GreedyRandomSampler.cpp [hunk 5]
```diff
@@ -145,7 +145,7 @@ GreedyRandomSampler::constructParameterPoints()
     if (d_use_latin_hypercube)
     {
         std::vector<double> frequencies;
-        std::vector<std::vector<double>> point_coordinates;
+        std::vector<std::vector<double> > point_coordinates;
```
Comment (kevinhkhuynh) [inline lib/algo/greedy/GreedyRandomSampler.cpp:148]: SWIG does not allow >>, it requires a space in between, so I updated these to > >

File: docs/Doxyfile [hunk 6]
```diff
@@ -993,7 +993,7 @@ RECURSIVE              = YES
 # Note that relative paths are relative to the directory from which doxygen is
 # run.
 
-EXCLUDE                =
+EXCLUDE                = lib/librom_wrap.cxx
```
Comment (kevinhkhuynh) [inline docs/Doxyfile:996]: We don't want to generate documentation for this SWIG file, so let's exclude it.

**Comment by chldkdtn** [comment]
I am pinging @pranav-vempati here because it is most likely that Pranav will pick this up and finish it.

File: lib/librom.i [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+%ignore SerialSVDDecomposition;
+%ignore SymmetricRightEigenSolve;
+%ignore NonSymmetricRightEigenSolve;
+%ignore SpaceTimeProduct;
+%include "linalg/Matrix.h"
+%include "linalg/Options.h"
+%include "linalg/BasisReader.h"
+%include "linalg/BasisGenerator.h"
+%include "algo/ParametricDMD.h"
+%apply (double* IN_ARRAY1, int DIM1) {(double* u_in, int /*unused*/)};
```
Comment (kevinhkhuynh) [inline lib/librom.i:64]: This allows us to pass in a numpy array in Python as a double array in C++

File: scripts/compile.sh [hunk 8]
```diff
@@ -96,4 +102,18 @@ elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
         -DUSE_MFEM=${USE_MFEM}
   make -j8
 fi
+
 popd
+
+if [[ $INSTALL_PYTHON == "true" ]]; then
+    . ${REPO_PREFIX}/scripts/python_setup.sh
+    cd ${REPO_PREFIX}/lib
+    ${REPO_PREFIX}/dependencies/swig/swig_install/bin/swig -I${REPO_PREFIX}/dependencies/swig/swig_install/share/swig/4.0.2 -I${REPO_PREFIX}/dependencies/swig/swig_install/share/swig/4.0.2/python -I${REPO_PREFIX}/dependencies/swig/swig_install/share/swig/4.0.2/std -c++ -python -py3 librom.i
```
Comment (kevinhkhuynh) [inline scripts/compile.sh:111]: This line creates a SWIG wrap py file.

File: scripts/compile.sh (hunk 2 above)
Comment (kevinhkhuynh) [reply scripts/compile.sh:117]: This line compiles the SWIG librom wrap file.

File: scripts/compile.sh (hunk 3 above)
Comment (kevinhkhuynh) [reply scripts/compile.sh:118]: This line converts the libROM wrap .o file into a .so Python library that is importable.

File: scripts/python_setup.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+LIB_DIR=$SCRIPT_DIR/../dependencies
+mkdir -p $LIB_DIR
+
+# Install SWIG-4.0.2
+cd $LIB_DIR
+if [ ! -d "swig" ]; then
+    wget http://prdownloads.sourceforge.net/swig/swig-4.0.2.tar.gz
+    tar -xvf swig-4.0.2.tar.gz
+    mv swig-4.0.2 swig
```
Comment (kevinhkhuynh) [inline scripts/python_setup.sh:28]: Here we download SWIG 4.0.2 and install it.

**Comment by kevinhkhuynh** [comment]
Useful documentation: https://www.swig.org/Doc1.3/Python.html

File: lib/librom.i (hunk 4 above)
Comment (kevinhkhuynh) [reply lib/librom.i:26]: DifferentialEvolution must also be eventually added, but it's not a big deal for now.

**Comment by chldkdtn** [comment]
@sullan2 Not sure how much this would help you for python interface task you are working on right now, but please check this PR out.

**Comment by dreamer2368** [comment]
I agree with you, @chldkdtn .

**Comment by kevinhkhuynh** [comment]
@chldkdtn This makes sense to me too.

**Pull request closed without merging**